File: relax-business-component/src/main/java/com/ustory/relax_business_component/plugin/PluginManager.java
package com.ustory.relax_business_component.plugin;

import android.content.Context;
import android.content.pm.ActivityInfo;
import android.content.res.AssetManager;
import android.content.res.Resources;

import java.lang.reflect.Field;
import java.util.HashMap;

import dalvik.system.DexClassLoader;

/**
 * Created by qiyue on 2018/1/30.
 */

public class PluginManager {

    private String dexOutputPath;
    private String mNativeLibDir;
    private Context mContext;
    private static PluginManager sInstance;

    private HashMap<String,Loader> plugins = new HashMap<>();

    public static PluginManager getInstance(Context context) {
        if (sInstance == null) {
            synchronized (PluginManager.class) {
                if (sInstance == null) {
                    sInstance = new PluginManager(context);
                }
            }
        }
        return sInstance;
    }

    private PluginManager(Context context) {
        mContext = context.getApplicationContext();
//        mNativeLibDir = mContext.getDir("pluginlib", Context.MODE_PRIVATE).getAbsolutePath();

    }

    public void loadApk(Plugin plugin){
        Loader loader = new Loader();

        if(checkNewVersion(plugin)){

        }else{
            /**
             * 最好开线程
             */
            int versionCode = loader.loadDefaultPlugin(mContext,plugin.getFileName());
            plugin.setVersion(versionCode);
            plugins.put(plugin.getKey(),loader);
        }

    }

    private boolean checkNewVersion(Plugin plugin) {

        //联网检查是否有新版本 ，如果有就下载，可以在下载到指定目录，然后调用 loader.loadPlugin
        return false;
    }

    public void replaceContextResources(Context context,Resources mResources){
        try {
            Field field = context.getClass().getDeclaredField("mResources");
            field.set(context, mResources);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }


    public HashMap<String, Loader> getPlugins() {
        return plugins;
    }

    public DexClassLoader getClassLoader(String key) {

        if (plugins.containsKey(key)){
            return plugins.get(key).getDexClassLoader();
        }
        return null;
    }

    public Resources getPluginResources(String key) {
        if (plugins.containsKey(key)){
            return plugins.get(key).getResources();
        }
        return null;
    }

    public AssetManager getAssets(String key) {
        if (plugins.containsKey(key)){
            return plugins.get(key).getAssets();
        }
        return null;
    }



    public ActivityInfo getActivityInfo(String key) {
        if (plugins.containsKey(key)){
            return plugins.get(key).getActivityInfo();
        }
        return null;

    }
}

File: relax-business-component/src/main/java/com/ustory/relax_business_component/plugin/PluginManager.kt
package com.ustory.relax_business_component.plugin

import android.content.Context
import android.content.pm.ActivityInfo
import android.content.res.AssetManager
import android.content.res.Resources
import dalvik.system.DexClassLoader
import java.util.*

/**
 * Created by qiyue on 2018/1/30.
 */

class PluginManager private constructor(context: Context) {

    private val dexOutputPath: String? = null
    private val mNativeLibDir: String? = null
    private val mContext: Context

    val plugins = HashMap<String?, Loader>()

    init {
        mContext = context.applicationContext
        //        mNativeLibDir = mContext.getDir("pluginlib", Context.MODE_PRIVATE).getAbsolutePath();

    }

    fun loadApk(plugin: Plugin) {
        val loader = Loader()

        if (checkNewVersion(plugin)) {

        } else {
            /**
             * 最好开线程
             */
            val versionCode = loader.loadDefaultPlugin(mContext, plugin.fileName!!)
            plugin.version = versionCode
            plugins[plugin.key] = loader
        }

    }

    private fun checkNewVersion(plugin: Plugin): Boolean {

        //联网检查是否有新版本 ，如果有就下载，可以在下载到指定目录，然后调用 loader.loadPlugin
        return false
    }

    fun replaceContextResources(context: Context, mResources: Resources) {
        try {
            val field = context.javaClass.getDeclaredField("mResources")
            field.set(context, mResources)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    fun getClassLoader(key: String): DexClassLoader? {

        return if (plugins.containsKey(key)) {
            plugins[key]?.dexClassLoader
        } else null
    }

    fun getPluginResources(key: String): Resources? {
        return if (plugins.containsKey(key)) {
            plugins[key]?.resources
        } else null
    }

    fun getAssets(key: String): AssetManager? {
        return if (plugins.containsKey(key)) {
            plugins[key]?.assets
        } else null
    }


    fun getActivityInfo(key: String): ActivityInfo? {
        return if (plugins.containsKey(key)) {
            plugins[key]?.activityInfo
        } else null

    }

    companion object {
        private var sInstance: PluginManager? = null

        fun getInstance(context: Context): PluginManager? {
            if (sInstance == null) {
                synchronized(PluginManager::class.java) {
                    if (sInstance == null) {
                        sInstance = PluginManager(context)
                    }
                }
            }
            return sInstance
        }
    }
}
